### ISSUE TITLE
Is the stable partition algorithm described in the paper actually O(1) space complexity?

### ISSUE THREAD

**Issue by rithik25** [schaetzc/stablepartition#2]
After reading the paper [Stable Minimum Space Partitioning in Linear Time](http://hjemmesider.diku.dk/~jyrki/Paper/KP1992bJ.pdf), I am not really convinced that the algorithm is actually O(1) space complexity.

Although the algorithm uses constant memory (technically it is O(log n) bits, but we can consider it O(1) in practical use), it doesn't use O(1) counters. The authors define O(1) space as "we assume that a constant number of extra storage locations, each capable for storing a word of O(log2 n) bits, is available" (page 2).  But later in algorithm D, the authors clearly use O(log(n)/log(log(n))) counters thereby contradicting their earlier statement. 

Even if we ignore the problem of counters, the algorithm most probably wouldn't work out that well in practical use. The algorithm can only have O(log2 n) extra space if we use a variable number of integer counters that are of variable size. The only way we can have a variable number of integer counters of variable size is if we dynamically allocate them and that would drastically slow down the algorithm, hence making the algorithm impractical.

Are my concerns justified or am I somehow misunderstanding the paper?

**Comment by schaetzc** [comment]
> Is [it] actually O(1) space

Yes, as far as I (and all the reviewers of the published paper) could tell.

You are right that Algorithm D uses a variable number of counters. However, these counters have a very limited range. To address an element inside a block with lg n elements, we only need log₂ lg n bits while a word has at least  log₂n bits. Therefore, **multiple counters can be encoded in a single word** using bit shift operations. When encoding the counters densely, we find ...

O(log n / log log n) counters using O(log log n) bits each, require a total of   
O( (log n / log log n) · (log log n) ) bits = O(log n) bits = O(1) words

Let's make this a bit more precise. From the paper we know ...

> Algorithm C sorts stably a bit-array of size m in O(m) time with lg m + m / lg m booleans and lg m + m / lg m + O(1) counters, each requiring at most ⌈log₂(m+1)⌉ bits.

We focus on the non-constant parts (lg m + m / lg m) only.  
Algorithm D applies C to each block of size lg n. So we plug m = lg n in above equation and get  
(lg lg n + lg n / lg lg n) booleans + (lg lg n + lg n / lg lg n) counters with at most ⌈log₂(lg n + 1)⌉ bits

Our words have log₂n bits.
To encode the booleans, we need ⌈(lg lg n + lg n / lg lg n) / log₂n⌉ words, [which is in O(1)](https://www.wolframalpha.com/input/?i=limit+%28log+log+n+%2B+log+n+%2F+log+log+n%29+%2F+log+n+as+n-%3Einfinity).
To encode the counters, we need  ⌈⌈log₂(lg n + 1)⌉ / log₂n⌉ words, [which is in O(1)](https://www.wolframalpha.com/input/?i=limit+log%28log+n+%2B+1%29+%2F+log+n+as+n-%3Einfinity).

**Comment by schaetzc** [comment]
I don't understand you second concern.

> can only have O(log2 n) extra space

Why? If you wanted you could use more, even though that defeats the whole point of this algorithm.

> The only way we can have a variable number of integer counters of variable size is if we dynamically allocate

Yes, that is right.

> and that would drastically slow down the algorithm, hence making the algorithm impractical.

Not really. There are many efficient algorithms that are not O(1) space.  
Also, this algorithm isn't very practical to begin with. In most real-world scenarios you'd probably ditch the O(1) space requirement in exchange for a more efficient out-of-place approach. Even though this algorithm is still in O(n) time, the trivial out-of-place O(n) time algorithm would probably beat it by magnitudes.

**Comment by rithik25** [comment]
> You are right that Algorithm D uses a variable number of counters. However, these counters have a very limited range. To address an element inside a block with lg n elements, we only need log₂ lg n bits while a word has at least log₂n bits. Therefore, **multiple counters can be encoded in a single word** using bit shift operations. When encoding the counters densely, we find ...

But before encoding the counters to one word, we needed O(log(n)/log(log(n))) counters right? Doesn't that itself go against the definition of O(1) space given in the paper ("we assume that a constant number of extra storage locations, each capable for storing a word of O(log2 n) bits, is available") since the number of counters is clearly not constant?

**Comment by rithik25** [comment]
> Why? If you wanted you could use more, even though that defeats the whole point of this algorithm.

I meant that without using a variable number of integer counters that are of variable size we wouldn't be able to achieve the O(log n) space complexity. Obviously smaller space complexity is better.

> > and that would drastically slow down the algorithm, hence making the algorithm impractical.

> Not really. There are many efficient algorithms that are not O(1) space.
Also, this algorithm isn't very practical to begin with. In most real-world scenarios you'd probably ditch the O(1) space requirement in exchange for a more efficient out-of-place approach. Even though this algorithm is still in O(n) time, the trivial out-of-place O(n) time algorithm would probably beat it by magnitudes.


I know that not having O(1) space complexity wouldn't really affect the speed. Here I meant that using dynamically allocated variables instead of static ones would really slow down the algorithm and make it useless in comparison to something like merge sort.

**Comment by schaetzc** [comment]
> But before encoding the counters to one word, we needed O(log(n)/log(log(n))) counters right?

No. The counters can be encoded and decoded one at a time. We only use a fixed amount of decoded counters at a time.

In pseudocode: You will never see something like `word[] allCounters` in the whole program. Instead of using `allCounters[i]` we simply use `decodeCounter(i)`.

An example of this can be seen in [`Partition.permute()`](https://github.com/schaetzc/stablepartition/blob/main/src/stablepartition/Partition.java#L112) where we only have the variables `i` and `k` for decoded counters.

**Comment by schaetzc** [comment]
I still don't get your second concern.

> without using a variable number of integer counters that are of variable size we wouldn't be able to achieve the O(log n) space complexity

First of all, we are already using a variable number of counters (wasn't this the main point of your first concern?). I don't know what you mean by *"variable size"*, but they certainly have different sizes depending on the block size. The more blocks we have, the smaller each counter is. This is the trick to getting everything into O(1) space.

Also, every algorithm in O(1) space is also in O(log n) space. Even if you didn't use any variable at all, you still would be in O(log n) space. Why do you even want to be in O(log n) space if the whole point of this algorithm is to be in-place?  

<sub>Maybe there is a confusion between bits and words? Space complexity measures the number of words. O(log n) bits have O(1) space complexity. But even then I don't understand what the point here is.</sub>  


> using dynamically allocated variables instead of static ones would really slow down the algorithm

To me, the point of in-place algorithms is to allow for huge inputs where you simply wouldn't have enough space (be it memory or storage) to process the input otherwise. That you might save some time on memory management is normally not of concern. Typically, in-place algorithms are even slower than out-of-place algorithms (this is known as the space-time tradeoff).

Either way, I think you are overestimating the cost of dynamic allocation by magnitudes. But if you really wanted, you could calculate an upper bound for the number of counters and allocate a fixed amount of memory beforehand. This is possible because you know that n ≤ 2 <sup>bits per word</sup>.

**Comment by rithik25** [comment]
> No. The counters can be encoded and decoded one at a time. We only use a fixed amount of decoded counters at a time.
> 
> In pseudocode: You will never see something like `word[] allCounters` in the whole program. Instead of using `allCounters[i]` we simply use `decodeCounter(i)`.
> 
> An example of this can be seen in [`Partition.permute()`](https://github.com/schaetzc/stablepartition/blob/main/src/stablepartition/Partition.java#L112) where we only have the variables `i` and `k` for decoded counters.

This response makes things much more clear to me. I now understand how the algorithm works out to be O(1) space. Thank you for taking out the time to respond to me.

**Comment by rithik25** [comment]
I would like to try helping out in completing this implementation of the algorithm. I would appreciate it if you could include me as a contributor.

**Comment by schaetzc** [comment]
Thank you for helping out. For starters I'd prefer a pull request first.
